util/confutil/config.go

```diff
@@ -1,39 +1,144 @@
-	"github.com/sirupsen/logrus"
+	fs "github.com/tonistiigi/fsutil/copy"
 )
 
-// ConfigDir will look for correct configuration store path;
-// if `$BUILDX_CONFIG` is set - use it, otherwise use parent directory
-// of Docker config file (i.e. `${DOCKER_CONFIG}/buildx`)
-func ConfigDir(dockerCli command.Cli) string {
-	if buildxConfig := os.Getenv("BUILDX_CONFIG"); buildxConfig != "" {
-		logrus.Debugf("using config store %q based in \"$BUILDX_CONFIG\" environment variable", buildxConfig)
-		return buildxConfig
+const defaultBuildKitConfigFile = "buildkitd.default.toml"
+
+type Config struct {
+	dir     string
+	chowner *osutil.User
+}
+
+type ConfigOption func(*configOptions)
+
+type configOptions struct {
+	dir string
+}
+
+func WithDir(dir string) ConfigOption {
+	return func(o *configOptions) {
+		o.dir = dir
+	}
+}
+
+func NewConfig(dockerCli command.Cli, opts ...ConfigOption) *Config {
+	co := configOptions{}
+	for _, opt := range opts {
+		opt(&co)
+	}
+
+	configDir := co.dir
+	if configDir == "" {
+		configDir = os.Getenv("BUILDX_CONFIG")
+		if configDir == "" {
+			configDir = filepath.Join(filepath.Dir(dockerCli.ConfigFile().Filename), "buildx")
+		}
+	}
+
+	var usr *osutil.User
+	home, _ := os.UserHomeDir()
+	if ok, _ := osutil.IsSubPath(home, configDir); ok {
+		usr = osutil.Sudoer()
 	}
 
-	buildxConfig := filepath.Join(filepath.Dir(dockerCli.ConfigFile().Filename), "buildx")
-	logrus.Debugf("using default config store %q", buildxConfig)
-	return buildxConfig
+	return &Config{
+		dir:     configDir,
+		chowner: usr,
+	}
+}
+
+// Dir will look for correct configuration store path;
+// if `$BUILDX_CONFIG` is set - use it, otherwise use parent directory
+// of Docker config file (i.e. `${DOCKER_CONFIG}/buildx`)
+func (c *Config) Dir() string {
+	return c.dir
 }
 
-// DefaultConfigFile returns the default BuildKit configuration file path
-func DefaultConfigFile(dockerCli command.Cli) (string, bool) {
-	f := path.Join(ConfigDir(dockerCli), "buildkitd.default.toml")
+// BuildKitConfigFile returns the default BuildKit configuration file path
+func (c *Config) BuildKitConfigFile() (string, bool) {
+	f := filepath.Join(c.dir, defaultBuildKitConfigFile)
 	if _, err := os.Stat(f); err == nil {
 		return f, true
 	}
 	return "", false
 }
 
+// MkdirAll creates a directory and all necessary parents within the config dir
+func (c *Config) MkdirAll(dir string, perm os.FileMode) error {
+	var chown fs.Chowner
+	if c.chowner != nil {
+		chown = func(user *fs.User) (*fs.User, error) {
+			return &fs.User{UID: c.chowner.UID, GID: c.chowner.GID}, nil
+		}
+	}
+	d := filepath.Join(c.dir, dir)
+	if err := fs.MkdirAll(d, perm, chown, nil); err != nil {
+		return err
+	}
+	if c.chowner == nil {
+		return nil
+	}
+	st, err := os.Stat(d)
+	if err != nil {
+		return err
+	}
+	downer := osutil.FileOwner(st)
+	if downer != nil && (downer.UID != c.chowner.UID || downer.GID != c.chowner.GID) {
+		return os.Chown(d, c.chowner.UID, c.chowner.GID)
+	}
```

Comment:
@tonistiigi I see you have a pending change that seems exactly what I need https://github.com/tonistiigi/fsutil/pull/214 but for ownership instead of timestamp :eyes:

Reply:
For this change if you want to use `fsutil` implementation, can't you just pass a chowner? Why do you need to get a list of created dirs?

Reply:
Ah, you already did that part. This question was about fixing old broken permissions? You can't do that with the `fsutil` implementation as well, as it only returns the directories it created. Not old directories.

Reply:
True it makes sense in fsutil only for newly created dirs. I keep this logic then.